### PR TITLE
feat: CLUE support for URL query parameter + rewrite hash parameters

### PR DESCRIPTION
**Pull request by thinkh**
Closes https://github.com/Caleydo/tdp_bi_bioinfodb/issues/1354

### Summary

* Adds two new ATDPApplication config flags:
   * `cluePropertyHandler`: determines if the query `?` or hash `#` should be used for CLUE parameters
   * `clueRewriteOtherProperty`: defines if the respective other property should be rewritten to maintain compatibility with existing URLs
* Refactor and restructured the `HashProperties` and `PropertyHandler`
* Refactor how the public session link is generated in `EditProvenanceGraphMenu.ts`

### Screenshots

[ordino-rewrite-clue-url.webm](https://user-images.githubusercontent.com/5851088/182543839-8157203e-e5d2-4e62-b327-36b2172e4947.webm)

See also the unit tests for the URL rewrite in [CLUEGraphManager.test.ts](https://github.com/datavisyn/tdp_core/pull/762/files#diff-9668eba1a8b06b07ddcbbb51df0632e4cf5d85066dc90e4a46381d1de7547346).

### How to use

In order to use the query parameter the following option in ATDPApplication must be set (default is `hash`):

```
const _ = new Ordino({
  cluePropertyHandler: 'query',
});
```

The rewrite is enabled with the flag `clueRewriteOtherProperty: true`.

### CODE REVIEW

**Comment by thinkh** [comment]
@puehringer Please review the code. If you like you can also test it in one of your apps, but the default behavior hasn't changed.

@oltionchampari Please test the functionality with Ordino in combination with https://github.com/Caleydo/ordino/pull/419 and https://github.com/Caleydo/ordino_public/pull/142.

File: src/ATDPApplication.ts [hunk 1]
```diff
@@ -74,14 +74,37 @@ export interface ITDPOptions {
...
   enableProvenanceUrlTracking?: boolean;
 
+  /**
+   * The CLUE parameters can be encoded in the hash '#clue_graph=...' (value: 'hash')
+   * or as query parameters `?clue_graph=...` (value: 'query') in the URL.
+   *
+   * @related clueRewriteOtherProperty
+   * @default hash
+   */
+  cluePropertyHandler?: 'hash';
```
Comment (oltionchampari) [inline src/ATDPApplication.ts:88]:
```suggestion
  cluePropertyHandler?: 'hash' | 'query';
```

File: src/ATDPApplication.ts [hunk 2]
```diff
@@ -74,14 +74,37 @@ export interface ITDPOptions {
...
+   *
+   * - With `cluePropertyHandler: 'hash'` it rewrites URLs with `?clue_graph=...` to `#clue_graph=...`
+   * - With `cluePropertyHandler: 'query'` it rewrites URLs with `#clue_graph=...` to `?clue_graph=...`
+   *
+   * If this flag is set to `false` the rewrite is disabled.
+   *
+   * @related cluePropertyHandler
+   * @default false
+   */
+  clueRewriteOtherProperty?: false;
```
Comment (oltionchampari) [inline src/ATDPApplication.ts:101]:
```suggestion
  clueRewriteOtherProperty?: boolean;
```

**Review by oltionchampari — CHANGES_REQUESTED** [review]
I noticed two minor things in the compiler. Did not look at the rest of the code

File: src/ATDPApplication.ts (hunk 1 above)
Comment (thinkh) [reply src/ATDPApplication.ts]: Thanks for pointing out. Fixed.

File: src/ATDPApplication.ts (hunk 2 above)
Comment (thinkh) [reply src/ATDPApplication.ts]: Thanks for pointing out. Fixed.

**Review by puehringer — APPROVED** [review]

**Comment by thinkh** [comment]
@oltionchampari I addressed all your points. Could you please have another look at the PRs and approve them so that we can proceed with the merge?

**Review by oltionchampari — APPROVED** [review]
Everything in order now

**Comment by oltionchampari** [comment]
Had a conversation with @thinkh and decided to merge

**Pull request merged**